### ISSUE TITLE
Possible bug: qwen_dna_collate_fn default silently disables SFT loss in train_dna_qwen.py

### ISSUE THREAD

**Issue by ShiLong-CN** [bowang-lab/BioReason#23]
Hi BioReason team — thanks for releasing this codebase, it's been a great starting point for our work.

While working on top of it we noticed what we believe is a subtle bug in `train_dna_qwen.py` that silently disables SFT supervision on both training and validation. Filing this as an issue first (rather than a PR) so you can confirm whether our reading is correct before anything is changed. Happy to follow up with a PR if you agree.

## Observed behavior

When training with `train_dna_qwen.py` (any `dataset_type`), the SFT loss curve behaves as if the model is receiving almost no supervision — train/val loss is degenerate from the first steps. Our internal runs showed this and after digging in we traced it to the collate function default, not the model or the data.

## Root cause (we think)

`qwen_dna_collate_fn` is defined in `bioreason/dataset/kegg.py` with:

```python
def qwen_dna_collate_fn(
    examples, processor, max_length_text, max_length_dna,
    return_answer_in_batch: bool = False,
    truncate_for_generation: bool = True,   # <-- default True
) -> Dict:
```
(see `bioreason/dataset/kegg.py:92-98`)

When `truncate_for_generation=True`, the collator (`kegg.py:206-253`) rebuilds `input_ids` / `attention_mask` / `labels` to only include tokens **up to and including** the `<|im_end|>\n<|im_start|>assistant\n` marker, and left-pads the rest. That is exactly what you want for generation batches, but it also drops the whole assistant answer span from `input_ids` and — crucially — from `labels`.

`train_dna_qwen.py` builds train / val `DataLoader`s with `partial(qwen_dna_collate_fn, ...)` at four sites, none of which override that default:

- `train_dataloader()` kegg branch (around `train_dna_qwen.py:405`)
- `train_dataloader()` `variant_effect_coding` branch (around `train_dna_qwen.py:437`)
- `train_dataloader()` `variant_effect_non_snv` branch (around `train_dna_qwen.py:470`)
- `val_dataloader()` shared collate (around `train_dna_qwen.py:559`)

So the batches that reach `_step` go through the generation-oriented truncation path. `_step` then runs the standard SFT loss:

```python
outputs = self.model(input_ids=..., attention_mask=..., labels=labels)
loss = outputs.loss
```
(`train_dna_qwen.py:168-177`)

But after truncation, the positions that should hold real label IDs for the answer have been dropped along with the tokens, so `labels` ends up almost entirely `-100`. The returned `outputs.loss` is computed over a near-empty supervision target, which is why train/val loss looks degenerate.

## Blast radius

- **Train loss** (`training_step` → `_step`): affected.
- **Val loss** (`validation_step` → `_step`): affected.
- **Test**: `_step` short-circuits at the top with `loss=0` when `prefix == "test"` (`train_dna_qwen.py:155-156`), so the test path happens to be insulated from this specific issue.
- **Generation / evaluation paths** (the periodic sample generation inside `_step`, and `on_test_epoch_end`): **not affected**. They re-scan each row's `input_ids` for the assistant marker per-sample and slice `gen_input_ids = input_ids[..., start_idx : assistant_pos + marker_len]` locally, so they don't depend on the collator having already truncated.

So the bug is specific to loss computation in train/val, and generation still works.

## Suggested direction (not a patch)

A couple of options, in rough order of size of change:

1. Pass `truncate_for_generation=False` explicitly at the four `partial(qwen_dna_collate_fn, ...)` sites in `train_dna_qwen.py`. Smallest diff, only touches this file. Leaves the collator's default unchanged for any other caller that might legitimately want generation-shaped batches.
2. Flip the default of `qwen_dna_collate_fn` itself from `True` to `False`, and have generation-time callers opt in. Conceptually cleaner (defaults match the training use case), but a wider behavior change if the flag is used elsewhere.

We'd lean toward option 1 unless you have a reason to prefer (2).

## Other files we checked

- `train_dna_only.py` uses a different collator (`dna_collate_fn`) without this parameter — not affected.
- `train_grpo.py` uses its own batching logic — not affected.

## Happy to help

If you confirm this reading, we're glad to open a PR with the minimal version of option 1 (or whichever direction you prefer). We also have an internal run we can re-do side-by-side to show the loss curves before and after, if that's useful for confirming the symptom.

Thanks again for the work on BioReason — please let us know if we've misread anything here.
